src/python/deepgnn/graph_engine/snark/dispatcher.py

```diff
@@ -291,11 +321,53 @@ def join(self):
             flag, output = self.q_out.get()
             self.node_count += output["node_count"]
             self.edge_count += output["edge_count"]
+            self.node_type_num = max(self.node_type_num, output["node_type_num"])
+            self.edge_type_num = max(self.edge_type_num, output["edge_type_num"])
+            self.node_feature_num = max(
+                self.node_feature_num, output["node_feature_num"]
+            )
+            self.edge_feature_num = max(
+                self.edge_feature_num, output["edge_feature_num"]
+            )
             self.partitions.append(output["partition"])
 
             assert flag == FLAG_WORKER_FINISHED_PROCESSING
         self.q_out.close()
 
+        for p in self.partitions:
+            real_node_type_num = len(p["node_type_count"])
+            if real_node_type_num < self.node_type_num:
+                p["node_type_count"].extend(
```

Comment:
Where is a test for this code? We didn't need to verify graph with variable features/types per partition before, because we had global information in meta.json before.

Reply:
where are tests for variable number of features per partition? I am particularly interested in how do we handle feature data files when one partition doesn't have any features and another partition has some.

Reply:
I added them to e2e tests - I'll add another for p0 has 0 features p1 has more.  https://github.com/microsoft/DeepGNN/pull/45/files#diff-62446f47649f4c1cb720810e69bfbeb8269aa8d178c0bea0729bcb9edbe7ea6b:~:text=def%20test_multi_partition_metadata()%3A